pkg/platform/runtime/buildexpression/merge/merge.go

```diff
@@ -0,0 +1,132 @@
+package merge
+
+import (
+	"encoding/json"
+	"reflect"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/multilog"
+	bpModel "github.com/ActiveState/cli/pkg/platform/api/buildplanner/model"
+	"github.com/ActiveState/cli/pkg/platform/api/mono/mono_models"
+	"github.com/ActiveState/cli/pkg/platform/runtime/buildexpression"
+)
+
+type AutoMergeNotPossibleError struct{ error }
+
+func IsAutoMergeNotPossibleError(err error) bool {
+	return errs.Matches(err, &AutoMergeNotPossibleError{})
+}
+
+type MergeConflictsError struct{ error }
+
+func IsMergeConflictsError(err error) bool {
+	return errs.Matches(err, &MergeConflictsError{})
+}
+
+func Merge(exprA *buildexpression.BuildExpression, exprB *buildexpression.BuildExpression, strategies *mono_models.MergeStrategies) (*buildexpression.BuildExpression, error) {
+	if !isAutoMergePossible(exprA, exprB) {
+		return nil, &AutoMergeNotPossibleError{errs.New("Unable to merge buildexpressions")}
+	}
+	if len(strategies.Conflicts) > 0 {
+		return nil, &MergeConflictsError{errs.New("Unable to merge buildexpressions due to conflicting requirements")}
+	}
+
+	// Update build expression requirements with merge results.
+	for _, req := range strategies.OverwriteChanges {
+		var op bpModel.Operation
+		switch req.Operation {
+		case mono_models.CommitChangeEditableOperationAdded:
+			op = bpModel.OperationAdded
+		case mono_models.CommitChangeEditableOperationRemoved:
+			op = bpModel.OperationRemoved
+		case mono_models.CommitChangeEditableOperationUpdated:
+			op = bpModel.OperationUpdated
+		default:
+			return nil, errs.New("Unknown requirement operation: %s", op)
+		}
+
+		var versionRequirements []bpModel.VersionRequirement
+		for _, constraint := range req.VersionConstraints {
+			data, err := constraint.MarshalBinary()
+			if err != nil {
+				return nil, errs.Wrap(err, "Could not marshal requirement version constraints")
+			}
+			m := make(map[string]string)
+			err = json.Unmarshal(data, &m)
+			if err != nil {
+				return nil, errs.Wrap(err, "Could not unmarshal requirement version constraints")
+			}
+			versionRequirements = append(versionRequirements, m)
```

Comment:
This is a neat solution! But we may be able to keep it simple, the `version_constraint` field holds the already marshalled values, and according to dashboard we're safe to use this despite it being marked deprecated:

https://activestate.slack.com/archives/CKMUXF64Q/p1689281817627129

Reply:
Ah, sadly `bpModel.VersionRequirements` is a `map[string]string`, and `req.VersionConstraint` is just a string (e.g. "=1.2.3"). The build planner is expecting something like `{"comparator": "eq", "version": "1.2.3"}`, so extracting the data from `req.VersionConstraints` (which can be tricked into to a `map[string]string`) is easier than interpreting `req.VersionConstraint` and creating a map from it.